extension/js/common/message-renderer.ts

```diff
@@ -0,0 +1,782 @@
+    // eslint-disable-next-line no-null/no-null
+    if (verifyRes.match === null || !Value.arr.hasIntersection(verifyRes.signerLongids, verifyRes.suppliedLongids)) {
+      MessageRenderer.renderMissingPubkey(renderModule, verifyRes.signerLongids[0]);
+    } else {
+      MessageRenderer.renderBadSignature(renderModule);
+    }
+  };
+
+  private static renderMissingPubkey = (renderModule: RenderInterfaceBase, signerLongid: string) => {
+    renderModule.renderSignatureStatus(`could not verify signature: missing pubkey ${signerLongid}`);
+  };
+
+  private static renderBadSignature = (renderModule: RenderInterfaceBase) => {
+    renderModule.renderSignatureStatus('bad signature');
+    renderModule.setFrameColor('red'); // todo: in what other cases should we set the frame red?
+  };
+
+  private static getVerificationPubs = async (signerEmail: string) => {
+    const email = Str.parseEmail(signerEmail).email;
+    if (!email) return [];
+    const parsedPubs = (await ContactStore.getOneWithAllPubkeys(undefined, email))?.sortedPubkeys ?? [];
+    // todo: we don't actually need parsed pubs here because we're going to pass them to the background page
+    // maybe we can have a method in ContactStore to extract armored keys
+    return parsedPubs.map(key => KeyUtil.armor(key.pubkey));
+  };
+
+  private static handlePrivateKeyMismatch = async (renderModule: RenderInterface, armoredPubs: string[], message: Uint8Array | string, isPwdMsg: boolean) => {
+    // todo - make it work for multiple stored keys
+    const msgDiagnosis = await MsgUtil.diagnosePubkeys({ armoredPubs, message });
+    if (msgDiagnosis.found_match) {
+      renderModule.renderErr(Lang.pgpBlock.cantOpen + Lang.pgpBlock.encryptedCorrectlyFileBug, undefined);
+    } else if (isPwdMsg) {
+      renderModule.renderErr(
+        Lang.pgpBlock.pwdMsgOnlyReadableOnWeb + MessageRenderer.btnHtml('ask sender to re-send', 'gray2 short reply_pubkey_mismatch'),
+        undefined
+      );
+    } else {
+      const startText =
+        msgDiagnosis.receivers === 1
+          ? Lang.pgpBlock.cantOpen + Lang.pgpBlock.singleSender + Lang.pgpBlock.askResend
+          : Lang.pgpBlock.yourKeyCantOpenImportIfHave;
+      renderModule.renderErr(
+        startText +
+          MessageRenderer.btnHtml('import missing key', 'gray2 settings_add_key') +
+          '&nbsp; &nbsp;' +
+          MessageRenderer.btnHtml('ask sender to update', 'gray2 short reply_pubkey_mismatch') +
+          '&nbsp; &nbsp;' +
+          MessageRenderer.btnHtml('settings', 'gray2 settings_keyserver'),
+        undefined
+      );
+    }
+  };
+
+  private static btnHtml = (text: string, addClasses: string) => {
+    return `<button class="button long ${addClasses}" style="margin:30px 0;" target="cryptup">${text}</button>`;
+  };
+
+  public renderMsg = ({ senderEmail, blocks }: { blocks: MsgBlock[]; senderEmail?: string }, showOriginal: boolean) => {
```

Comment:
This class has the entry points for rendering to be re-used by both GmailElementReplacer and extensions's own inbox page.
Only a few methods are public:
- renderMsg
- processAttachment
- processInlineBlocks
- msgGetProcessed.

Some other methods may be grouped in separate classes, I believe.

Also, some methods from `Mime` class, I believe, should be moved here or into a separate static class, namely `processBody` and `processAttachments` (former `processDecoded`) as they go beyond Mime-related logic.

Reply:
> By the way, is there a lot of functionality left to be refactored in this PR?

No, all the current tests succeed, I'm looking at scenarios that weren't tested yet to see if anything is broken.
Mainly doing simplifications and adding tests.